### PR TITLE
feat: free tx db entry

### PR DESCRIPTION
**Pull request by mtwaro**

### CODE REVIEW

File: src/staking/free_tx_info.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    bool isValid() const { return m_valid; }
+    uint32_t getUsedLimit() const { return m_used_limit; }
+    uint32_t getLimit() const { return m_limit; }
+    uint32_t getCurrentWindowStartHeight() const { return m_current_window_start_height; }
+    const std::set<uint256>& getActiveStakeIds() const { return m_active_stake_ids; }
+
+    void setLimit(const uint32_t limit) { m_limit = limit; }
+    void setUsedLimit(const uint32_t used_limit) { m_used_limit = used_limit; }
+    void setActiveStakeIds(std::set<uint256> active_stake_ids) { m_active_stake_ids = std::move(active_stake_ids); }
```
Comment (Jakub123Jakub123) [inline src/staking/free_tx_info.h:37]: active_stake_ids is moved so it could be && reference

File: src/staking/stakes_db.cpp [hunk 2]
```diff
@@ -426,12 +444,71 @@ const AmountByPeriodArray& CStakesDBCache::getAmountsByPeriods() const {
...
     }
     m_stakes_map[entry.getKey()] = entry;
     return true;
-}
\ No newline at end of file
+}
+
+CFreeTxInfo CStakesDBCache::getFreeTxInfoForScript(const CScript &script) const {
+    if (m_viewonly) {
+        return m_base_db->getFreeTxInfoForScript(script);
```
Comment (Jakub123Jakub123) [inline src/staking/stakes_db.cpp:459]: I do not see big difference between this function and body of the getFreeTxInfoForScript. In ideal world this logic should be abstracted into some kind of container class. Or algorithm/template function.

File: src/staking/stakes_db.cpp [hunk 3]
```diff
@@ -234,6 +251,7 @@ CStakesDBCache::CStakesDBCache(CStakesDB* db, bool fViewOnly, size_t max_cache_s
         m_script_to_active_stakes.insert(m_base_db->getScriptMap().begin(), m_base_db->getScriptMap().end());
         m_best_block_hash = m_base_db->getBestBlock();
         m_amounts_by_periods = m_base_db->getAmountsByPeriods();
+        m_free_tx_info.insert(m_base_db->getFreeTxInfoMap().begin(), m_base_db->getFreeTxInfoMap().end());
```
[the remark this answers is not in the document]
Comment (mtwaro) [reply src/staking/stakes_db.cpp:254]: No performance or readability gain comes from this.

File: src/staking/free_tx_info.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <utility>
+#include <vector>
+#include <boost/serialization/access.hpp>
+#include <uint256.h>
+
+class CFreeTxInfo {
+    friend class boost::serialization::access;
+    template<typename Archive>
+    void serialize(Archive& archive, const unsigned int version) {
```
[the remark this answers is not in the document]
Comment (mtwaro) [reply src/staking/free_tx_info.h:12]: It's the implementation of a method called by `boost::serialization`, and it must match the declaration.

File: src/staking/stakes_db.cpp (hunk 2 above)
Comment (mtwaro) [reply src/staking/stakes_db.cpp:459]: I'll do a common interface for `CStakesDB` and `CStakesDBCache` if we have some time left :)

**Review by Jakub123Jakub123 — APPROVED** [review]

**Pull request merged**